### PR TITLE
[MINI-3419] Signature verification

### PR DESCRIPTION
**Pull request by Climbatize**
# Description
This PR implements files signature check in Mini App SDK. It is partly based on @imaginaris work in https://github.com/rakutentech/ios-signatureverifier, the SDK code has been migrated and softly refactored to better suit our needs.
Implementing signature check Pod automatically enables it in MiniApp SDK. But by default a signature check failure does not prevent the Mini App to be displayed, it just triggers an analytics event. An optional parameter has been added in the case we would want to prevent it.
 
A setting screen has been added to help testing signature check failure parameter.

Screen Shot 2021-09-08 at 9 34 36 
Screen Shot 2021-09-08 at 9 34 47

## Links
[MINI-3419](https://jira.rakuten-it.com/jira/browse/MINI-3419)

# Checklist
- [X] I have read the [contributing guidelines](CONTRIBUTING.md).
- [X] I have completed the [SDK Development Learning Path](CONTRIBUTING.md#SDK-Development-Learning-Path) (if submitting a feature PR)
- [X] I wrote/updated tests for new/changed code
- [X] I removed all sensitive data before every commit, including API endpoints and keys
- [X] I ran `fastlane ci` without errors

### CODE REVIEW

**Review by laubengaier — APPROVED** [review]
nice work, left a little regex improvement 😄

File: MiniApp/Classes/signature/Security/SignatureKeyStore.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                kSecAttrAccount as String: account,
+                kSecValueData as String: data,
+                kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
+            ]
+            status = SecItemAdd(queryAdd as CFDictionary, nil)
+        }
+
+        if status != errSecSuccess {
+            let error: String?
+            if #available(iOS 11.3, *) {
```
Comment (imaginaris) [inline MiniApp/Classes/signature/Security/SignatureKeyStore.swift:57]: I think this #if can be removed because your target version is iOS 13.0, right?

Comment (Climbatize) [reply MiniApp/Classes/signature/Security/SignatureKeyStore.swift:57]: Of course! Thanks for the heads up

File: MiniApp/Classes/core/Downloader/MiniAppClient.swift [hunk 2]
```diff
@@ -145,7 +161,22 @@ internal class MiniAppClient: NSObject, URLSessionDownloadDelegate {
...
+                    self?.delegate?.fileDownloaded(at: location, downloadedURL: destinationURL, signatureChecked: isVerified || !requireMiniAppSignatureVerification)
+                }
+            } else {
+                MiniAppAnalytics.sendAnalytics(event: .signatureFailure, miniAppId: ids.0, miniAppVersion: ids.1)
+                delegate?.fileDownloaded(at: location, downloadedURL: destinationURL, signatureChecked: !requireMiniAppSignatureVerification)
+            }
+        #else
+            MiniAppAnalytics.sendAnalytics(event: .signatureFailure, miniAppId: ids.0, miniAppVersion: ids.1)
+            delegate?.fileDownloaded(at: location, downloadedURL: destinationURL)
+        #endif
```
Comment (rleojoseph) [inline MiniApp/Classes/core/Downloader/MiniAppClient.swift:179]: Do you think we can movubg this to a separate method would be good idea?

File: MiniApp/Classes/core/MiniApp.swift [hunk 3]
```diff
@@ -1,7 +1,7 @@
 /// Mini App Public API methods
 public class MiniApp: NSObject {
     private static let shared = MiniApp()
-    private let realMiniApp = RealMiniApp()
+    let realMiniApp = RealMiniApp()
```
Comment (rleojoseph) [inline MiniApp/Classes/core/MiniApp.swift:4]: Any reason? I see the build getting through even with private.

**Review by rleojoseph — APPROVED** [review]
Overall, Looks good to me :+1:

Just a few comments.

File: MiniApp/Classes/core/MiniAppAnalytics.swift [hunk 4]
```diff
@@ -10,13 +10,16 @@ internal enum MiniAppRATEvent: String, CaseIterable {
...
 
     func name() -> String {
-        return "mini_app_\(rawValue)"
+        "mini_app_\(rawValue)"
     }
 
     func eType() -> MiniAppRATEventType {
         switch self {
+        case .signatureFailure:
+            return .custom
```
Comment (corycaywood) [inline MiniApp/Classes/core/MiniAppAnalytics.swift:22]: I don't think this is correct - we can only use pre-defined values for the `etype`. There aren't really any that make sense for this case, but maybe `async` is the closest? And then we would need to set another parameter to distinguish that this is the signature failure event. How about this?

```
etype = async
actype = signature_failure
```

File: Example/Controllers/Extensions/ViewController+MiniApp.swift [hunk 5]
```diff
@@ -112,6 +112,10 @@ extension ViewController: MiniAppNavigationDelegate {
         switch error {
         case .metaDataFailure:
             metaDataFailure()
+        case .invalidSignature:
+            self.displayAlert(title: MASDKLocale.localize("miniapp.sdk.ios.error.title"), message: MASDKLocale.localize(.signatureFailed), dismissController: true) { _ in
```
Comment (corycaywood) [inline Example/Controllers/Extensions/ViewController+MiniApp.swift:116]: Does App Center let you manually send a crash report/log? Maybe we could do that here so we have that extra information from the Demo App?

**Review by corycaywood — APPROVED** [review]
Nice work! 👍  Just see my comment about analytics.

File: MiniApp/Classes/core/MiniAppAnalytics.swift (hunk 4 above)
Comment (corycaywood) [reply MiniApp/Classes/core/MiniAppAnalytics.swift:22]: I just noticed that the following is what @khairul-alam-licon  used on Android, so probably this is okay:
```
etype = click
actype = mini_app_signature_validation_fail
```

File: MiniApp/Classes/core/Downloader/MiniAppClient.swift (hunk 2 above)
Comment (Climbatize) [reply MiniApp/Classes/core/Downloader/MiniAppClient.swift]: yeah, makes things cleaner

File: MiniApp/Classes/core/MiniApp.swift (hunk 3 above)
Comment (Climbatize) [reply MiniApp/Classes/core/MiniApp.swift:4]: it was for some tests I made and did not need in the end, I'll fix this

File: Example/Controllers/Extensions/ViewController+MiniApp.swift (hunk 5 above)
Comment (Climbatize) [reply Example/Controllers/Extensions/ViewController+MiniApp.swift:116]: Unfortunately App Center does not have this feature, we could have a dedicated slack channel for this kind of things

**Pull request merged**